scapy/layers/hsrp.py

```diff
@@ -19,71 +19,213 @@
+            lambda pkt: valid_net(pkt.virtualIP)
+        )
+    ]
+
+    def post_build(self, pkt, pay):
+        if self.ipVer is None:
+            ip_ver = 4
+
+            if valid_net6(self.virtualIP):
+                ip_ver = 6
+
+            pkt = pkt[:5] + struct.pack("B", ip_ver) + pkt[6:]
+
+        if self.len is None:
+            pkt = pkt[:1] + struct.pack("B", len(pkt) - 2) + pkt[2:]
+        return pkt + pay
+
+    def guess_payload_class(self, payload):
+        if len(payload) > 0:
+            hsrp_auth_payload_type = payload[0]
+            if hsrp_auth_payload_type == 3:
+                return HSRPv2TextAuth
+            elif hsrp_auth_payload_type == 4:
+                return HSRPmd5
+        return Packet.guess_payload_class(self, payload)
 
 
 class HSRPmd5(Packet):
+    """
+    MD5 Authentication header for HSRP (version 1 and 2).
+    """
     name = "HSRP MD5 Authentication"
     fields_desc = [
         ByteEnumField("type", 4, {4: "MD5 authentication"}),
         ByteField("len", None),
-        ByteEnumField("algo", 0, {1: "MD5"}),
+        ByteEnumField("algo", 1, {1: "MD5"}),
         ByteField("padding", 0x00),
         XShortField("flags", 0x00),
         SourceIPField("sourceip", None),
         XIntField("keyid", 0x00),
-        StrFixedLenField("authdigest", b"\00" * 16, 16)]
+        StrFixedLenField("authdigest", b"\x00" * 16, 16)]
 
-    def post_build(self, p, pay):
-        if self.len is None and pay:
-            tmp_len = len(pay)
-            p = p[:1] + hex(tmp_len)[30:] + p[30:]
-        return p
+    def post_build(self, pkt, pay):
+        if self.len is None:
+            pkt = pkt[:1] + struct.pack("B", len(pkt) - 2) + pkt[2:]
+        return pkt + pay
+
+
+class HSRPv2TextAuth(Packet):
+    """
+    Default plain text authentication header.
+    This is only used with HSRP version 2.
+    """
+    name = "HSRP Authentication"
+    fields_desc = [
+        ByteEnumField("type", 3, {3: "Text Authentication TLV"}),
+        ByteField("len", 8),
+        StrFixedLenField("auth", b"cisco" + b"\x00" * 3, 8)
+    ]
 
 
 bind_bottom_up(UDP, HSRP, dport=1985)
 bind_bottom_up(UDP, HSRP, sport=1985)
-bind_bottom_up(UDP, HSRP, dport=2029)
-bind_bottom_up(UDP, HSRP, sport=2029)
+bind_bottom_up(UDP, HSRPv2, dport=1985)
+bind_bottom_up(UDP, HSRPv2, sport=1985)
+bind_bottom_up(UDP, HSRPv2, dport=2029)
+bind_bottom_up(UDP, HSRPv2, sport=2029)
+
 bind_layers(UDP, HSRP, dport=1985, sport=1985)
-bind_layers(UDP, HSRP, dport=2029, sport=2029)
+bind_layers(UDP, HSRPv2, dport=1985, sport=1985)  # HSRP v2 with IPv4
+bind_layers(UDP, HSRPv2, dport=2029, sport=2029)  # HSRP v2 with IPv6
+
 DestIPField.bind_addr(UDP, "224.0.0.2", dport=1985)
+DestIPField.bind_addr(UDP, "224.0.0.102", dport=1985)
 DestIP6Field.bind_addr(UDP, "ff02::66", dport=2029)
+
+DestIP6Field.bind_addr(HSRPv2, "ff02::66")
+DestIPField.bind_addr(HSRPv2, "224.0.0.102")
+DestIPField.bind_addr(HSRP, "224.0.0.2")
```

Comment:
What are those for. (in which case do you have a `DestField` whose payload is HSRP?)

Reply:
This is incorrect code that was leftover after testing, I removed it.